src/hierarchical.rs

```diff
@@ -0,0 +1,493 @@
+use std::cmp::{Ordering, min, max};
+use std::f64::INFINITY;
+use std::mem::replace;
+
+use {Point};
+
+
+/// A function to compute the inter-cluster distance between set `A` and `B`.
+pub type LinkageFunction = Fn(&Fn(usize, usize) -> f64, &[usize], &[usize]) -> f64;
+
+
+/// Hierarchical linkage criteria.
+///
+/// The linkage decides how the distance between clusters is computed.
+#[derive(Copy, Clone)]
+pub enum LinkageCriterion<'a> {
```

Comment:
> which, being a simple distance calculation, could take a few nanoseconds to compute, so the overhead of the virtual call can be significant

The calls to `naive`'s closure was something like... 10%-15% of the total time for `complete`, so I do think this is a pretty large cost.